File: backend/app/schemas/jobseeker.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.employer import ChargeFlagsPayload


class JobseekerProfilePayload(BaseModel):
    id: UUID
    app_user_id: UUID
    auth_user_id: UUID
    full_name: str | None
    phone: str | None
    address: str | None
    city: str | None
    zip: str | None
    transit_type: Literal["own_car", "public_transit", "both"] | None
    charges: ChargeFlagsPayload
    profile_complete: bool
    status: Literal["active", "hired"]
    created_at: datetime | None
    updated_at: datetime | None


class JobseekerProfileResponse(BaseModel):
    profile: JobseekerProfilePayload


class JobseekerProfileUpdateRequest(BaseModel):
    full_name: str | None = Field(default=None, min_length=1, max_length=255)
    phone: str | None = Field(default=None, max_length=32)
    address: str | None = Field(default=None, max_length=255)
    city: str | None = Field(default=None, max_length=128)
    zip: str | None = Field(default=None, max_length=16)
    transit_type: Literal["own_car", "public_transit", "both"] | None = None
    charges: ChargeFlagsPayload | None = None


class JobseekerProfileUpdateResultPayload(BaseModel):
    id: UUID
    profile_complete: bool
    updated_at: datetime | None


class JobseekerProfileUpdateResponse(BaseModel):
    profile: JobseekerProfileUpdateResultPayload


class JobseekerListItemPayload(BaseModel):
    id: UUID
    full_name: str | None
    city: str | None
    zip: str | None
    transit_type: Literal["own_car", "public_transit", "both"] | None
    status: Literal["active", "hired"]


class JobseekerApplicationSummaryPayload(BaseModel):
    id: UUID
    status: Literal["submitted", "reviewed", "hired"]
    job_listing_id: UUID


class AdminJobseekerDetailResponse(BaseModel):
    jobseeker: JobseekerProfilePayload
    applications: list[JobseekerApplicationSummaryPayload]


class AdminJobseekerUpdateRequest(JobseekerProfileUpdateRequest):
    status: Literal["active", "hired"] | None = None


class AdminJobseekerUpdateResponse(BaseModel):
    jobseeker: JobseekerProfileUpdateResultPayload

File: frontend/jobseeker/src/types/jobseeker.ts
export interface ChargeFlags {
  sex_offense: boolean
  violent: boolean
  armed: boolean
  children: boolean
  drug: boolean
  theft: boolean
}

export interface JobseekerProfile {
  id: string
  app_user_id: string
  auth_user_id: string
  full_name: string | null
  phone: string | null
  address: string | null
  city: string | null
  zip: string | null
  transit_type: 'own_car' | 'public_transit' | 'both' | null
  charges: ChargeFlags
  profile_complete: boolean
  status: 'active' | 'hired'
  created_at: string | null
  updated_at: string | null
}

export interface JobListing {
  id: string
  employer_id: string
  title: string
  description: string | null
  location_address: string | null
  city: string | null
  zip: string | null
  transit_required: 'own_car' | 'any'
  disqualifying_charges: ChargeFlags
  transit_accessible: boolean | null
  job_lat: number | null
  job_lon: number | null
  review_status: 'pending' | 'approved' | 'rejected'
  lifecycle_status: 'open' | 'closed'
  review_note: string | null
  reviewed_by: string | null
  reviewed_at: string | null
  created_at: string | null
  updated_at: string | null
}

export interface PaginatedResponse<T> {
  items: T[]
  meta: {
    page: number
    page_size: number
    total_items: number
    total_pages: number
  }
}

export interface JobListItem {
  job: JobListing
  is_eligible: boolean
  ineligibility_tag: string | null
}

export interface JobListFilters {
  page?: number
  search?: string
  city?: string
  transit_required?: 'own_car' | 'any' | ''
  is_eligible?: boolean
}

export interface JobDetailPayload {
  job: JobListing
  eligibility: {
    is_eligible: boolean
    ineligibility_tag: string | null
  }
}

export interface ApplicationListItem {
  id: string
  status: 'submitted' | 'reviewed' | 'hired'
  applied_at: string
  updated_at: string | null
  job: {
    id: string
    title: string
    city: string | null
    lifecycle_status: 'open' | 'closed'
  }
}

export interface ApplicationPayload {
  id: string
  jobseeker_id: string
  job_listing_id: string
  status: 'submitted' | 'reviewed' | 'hired'
  applied_at: string
  updated_at: string | null
}

export interface JobseekerProfileFormValues {
  full_name: string
  phone: string
  address: string
  city: string
  zip: string
  transit_type: 'own_car' | 'public_transit' | 'both' | ''
  charges: ChargeFlags
}
